### DISCUSSION TITLE
Is the C API for skia still intended to be stable?

### DISCUSSION THREAD

**Discussion by Mis012** [mono/SkiaSharp#2585] [Q&A]
I see that this project is keeping alive the C API for Skia, which Google has sadly removed from the upstream.
I have the following questions:
1. why is this not upstream? Google claimed that there was no interest in maintaining the C API, but clearly you are doing just that
2. The C API was intended to be stable (at least "since 1.0.0", but the version of the C API is not exactly written anywhere so who knows whether that has been reached yet). Is this still the goal?

**Comment by mattleibow** [comment]
The C API is stable in that between milestones it is additive. But, it really is at the whim of the underlying C++ API.

This is not upstream because Google is not super interested in it and I am just using it as an interop layer. So it is not always so nice to use.

I use it from C# and that is where I try and keep it stable. However, I am happy to take PRs to make the C API a more stable layer. Google does the C++ API, I do the C# API and whoever needs the C API can help make it more stable and nice to use.

**Reply by Mis012** [reply]
Well, I would happy to provide some PRs to make the API nicer to use and easier to keep stable, however the main thing is that whenever the API is changed in a way that e.g. necessitates adding parameters that were not being passed before (even in a different format which could be adapted), one would need to e.g call the new function `_v2` (or 3,4,...) and make the old function a wrapper which just passes whatever "defaults" to get the old behavior. Of course adapting the parameters when only the way they are passed changes could be done transparently with no `_v2` needed.  
Realistically, whenever you would now change the C API incompatibly, we would need to instead brainstorm how best to avoid doing that and make some sort of documentation so that consistency is ensured into the future.  
As someone who has been doing this for a while, I assume you might already have some idea about how the C++ API might change and what we should avoid in the C api in order to make it easier to adapt to these changes in the future without changing the C api itself.
